File: quicksort2.php
<?php

    function partition($ar) {
        $pivot = $ar[0];
        $left = $equal = $right = [];

        foreach ($ar as $elem) {
            if ($elem < $pivot) {
                array_push($left, $elem);
            } elseif ($elem == $pivot) {
                array_push($equal, $elem);
            } else {
                array_push($right, $elem);
            }
        }
        return [$left, $equal, $right];
    }

    function quickSort($ar) {
        if (count($ar) <= 1) {
            return $ar;
        }
        
        //get partiotions
        $parts = partition($ar);
        $left = $parts[0];
        $equal = $parts[1];
        $right = $parts[2];
        $merged = array_merge(quickSort($left), $equal, quickSort($right));
        echo implode(" ", $merged) . "\n";
        
        return $merged;
    }

    $fp = fopen("php://stdin", "r");

    fscanf($fp, "%d", $m);

    $ar = explode(' ', trim(fgets($fp)));
    
    quickSort($ar);

?>

File: quicksort2.py
def partition(arr):
    pivot = arr[0]
    left, equal, right = [], [], []
    for elem in arr:
        if elem < pivot:
            left.append(elem)
        elif elem == pivot:
            equal.append(elem)
        else:
            right.append(elem)

    return left, equal, right

def quicksort(arr):
    if len(arr) <= 1: 
        return arr
    
    left, equal, right = partition(arr)
    merged = quicksort(left) + equal + quicksort(right)
    print(" ".join(str(x) for x in merged))
    
    return merged


if __name__ == '__main__':
    n = input()

    ar = list(map(int,input().rstrip().split()))

    quicksort(ar)